File: JAVA/Recover Binary Search Tree.java
public class Solution {
    public void recoverTree(TreeNode root) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode n = root;
        TreeNode last = null;
        TreeNode current = null;
        TreeNode n1 = null;
        TreeNode n2 = null;
        while (n != null || !stack.empty()) {
            if (n != null) {
                stack.push(n);
                n = n.left;
            }
            else {
                n = stack.pop();
                // visit n
                last = current;
                current = n;
                if (last != null && current != null && last.val > current.val) {
                    if (n1 == null) {
                        n1 = last;
                        n2 = current;
                    }
                    else {
                        n2 = current;
                    }
                }
                n = n.right;
            }
        }
        if (n1 != null && n2 != null) {
            int tmp = n1.val;
            n1.val = n2.val;
            n2.val = tmp;
        }
    }
}

File: Recover Binary Search Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void recoverTree(TreeNode *root) {
        stack<TreeNode *> stk;
        TreeNode *node = root;
        TreeNode *prev = NULL;
        TreeNode *first = NULL;
        TreeNode *second = NULL;
        while (!stk.empty() || node != NULL) {
            if (node != NULL) {
                stk.push(node);
                node = node->left;
            } else {
                node = stk.top();
                stk.pop();
                // visit node
                if (prev != NULL) {
                    if (prev->val > node->val) {
                        if (first == NULL) {
                            first = prev;
                            second = node;
                        } else {
                            second = node;
                            break;
                        }
                    }
                }
                prev = node;
                node = node->right;
            }
        }
        if (first != NULL && second != NULL) {
            int tmp = first->val;
            first->val = second->val;
            second->val = tmp;
        }
    }
};
